concrete_types.hh

```diff
@@ -33,15 +33,39 @@
 
 struct empty_type_impl final : public abstract_type {
     empty_type_impl();
+
+    virtual bool contains_set_or_map() const {
+        return false;
+    }
+
+    virtual bool contains_collection() const {
```

Comment:
Just making sure: contains_colection() is almost the same as contains_set_or_map() but also checks for a list?

Reply:
It's better to add this as a visitor, we're trying to simplify abstract_type.

Reply:
(see types.cc for many visitors)

Reply:
> Just making sure: contains_colection() is almost the same as contains_set_or_map() but also checks for a list?

yes, added a blurb to make this more clear

> It's better to add this as a visitor, we're trying to simplify abstract_type.

:+1:, changed to visitor, there is much less code now